File: cli/internal/typer/generator/platforms/kotlin/testdata/validator/src/test/kotlin/EventValidationPlugin.kt
package com.rudderstack.ruddertyper

import com.rudderstack.sdk.kotlin.core.Analytics
import com.rudderstack.sdk.kotlin.core.internals.models.Event
import com.rudderstack.sdk.kotlin.core.internals.models.EventType
import com.rudderstack.sdk.kotlin.core.internals.models.*
import org.junit.jupiter.api.*
import com.rudderstack.sdk.kotlin.core.internals.plugins.Plugin
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject

private var emptyJsonObject = buildJsonObject {}

sealed class EventValidation(val type: EventType) {
    data class TrackValidation(
        val name: String,
        val properties: JsonObject = emptyJsonObject,
    ): EventValidation(type = EventType.Track)

    data class IdentifyValidation(
        val userId: String,
        val traits: JsonObject = emptyJsonObject
    ): EventValidation(type = EventType.Identify)

    data class ScreenValidation(
        val screenName: String,
        val properties: JsonObject = emptyJsonObject
    ): EventValidation(type = EventType.Screen)

    data class GroupValidation(
        val groupId: String,
        val traits: JsonObject = emptyJsonObject
    ): EventValidation(type = EventType.Group)
}

class EventValidationPlugin : Plugin {
    override val pluginType: Plugin.PluginType = Plugin.PluginType.OnProcess
    override lateinit var analytics: Analytics

    private var received: MutableList<Event> = mutableListOf()
    private var validationIndex: Int = 0

    override fun setup(analytics: Analytics) {
        super.setup(analytics)
    }

    override suspend fun intercept(event: Event): Event? {
        println("EventFilteringPlugin: Intercepting event: $event")
        received.add(event)
        return null
    }

    fun validateCount(count: Int, timeoutMs: Long = 5000) {
        val startTime = System.currentTimeMillis()
        val pollIntervalMs = 100L

        // Wait for events to arrive with timeout
        while (received.size < count) {
            if (System.currentTimeMillis() - startTime > timeoutMs) {
                throw AssertionError(
                    "Timeout waiting for events. " +
                    "Expected $count events, but only received ${received.size} after ${timeoutMs}ms"
                )
            }
            Thread.sleep(pollIntervalMs)
        }

        // Check if more events were received than expected
        if (received.size > count) {
            throw AssertionError(
                "Received more events than expected. " +
                "Expected $count events, but received ${received.size}. " +
                "Extra event: ${received[count]}"
            )
        }
    }

    public fun validateNext(expected: EventValidation) {
        validateEvent(received[validationIndex], expected)
        validationIndex++
    }

    private fun validateEvent(received: Event, expected: EventValidation) {
        when (received) {
            is TrackEvent -> {
                if (expected !is EventValidation.TrackValidation) {
                    throw AssertionError("Expected track event, but got ${expected.type}")
                }
                Assertions.assertEquals(expected.name, received.event, "Track event name mismatch")
                Assertions.assertEquals(expected.properties, received.properties, "Track event properties mismatch")
            }
            is IdentifyEvent -> {
                if (expected !is EventValidation.IdentifyValidation) {
                    throw AssertionError("Expected identify event, but got ${expected.type}")
                }
                Assertions.assertEquals(expected.userId, received.userId, "Identify userId mismatch")
                val traits = received.context["traits"]
                Assertions.assertEquals(expected.traits, traits, "Identify traits mismatch")
            }
            is ScreenEvent -> {
                if (expected !is EventValidation.ScreenValidation) {
                    throw AssertionError("Expected screen event, but got ${expected.type}")
                }
                Assertions.assertEquals(expected.screenName, received.screenName, "Screen event name mismatch")
                Assertions.assertEquals(expected.properties, received.properties, "Screen event properties mismatch")
            }
            is GroupEvent -> {
                if (expected !is EventValidation.GroupValidation) {
                    throw AssertionError("Expected group event, but got ${expected.type}")
                }
                Assertions.assertEquals(expected.groupId, received.groupId, "Group event groupId mismatch")
                Assertions.assertEquals(expected.traits, received.traits, "Group event traits mismatch")
            }
            else -> throw AssertionError("Unexpected event type: ${received.type}")
        }
    }
}

File: cli/internal/typer/generator/platforms/swift/testdata/validator/Sources/RudderTyperTests/EventValidationPlugin.swift
import Foundation
import XCTest
import RudderStackAnalytics

enum EventValidation {
    case track(name: String, properties: [String: Any] = [:])
    case identify(userId: String, traits: [String: Any] = [:])
    case screen(screenName: String, properties: [String: Any] = [:])
    case group(groupId: String, traits: [String: Any] = [:])
}

final class EventValidationPlugin: Plugin {
    var pluginType: PluginType = .onProcess
    var analytics: Analytics?

    private let lock = NSLock()
    private var received: [Event] = []
    private var validationIndex: Int = 0

    func setup(analytics: Analytics) {
        self.analytics = analytics
    }

    func teardown() {
        analytics = nil
    }

    func intercept(event: Event) -> Event? {
        lock.lock()
        received.append(event)
        lock.unlock()
        return nil
    }

    func validateCount(_ expected: Int, timeout: TimeInterval = 5.0, file: StaticString = #file, line: UInt = #line) {
        let deadline = Date().addingTimeInterval(timeout)
        while currentCount() < expected {
            if Date() >= deadline {
                XCTFail("Timeout waiting for events. Expected \(expected), got \(currentCount()) after \(timeout)s", file: file, line: line)
                return
            }
            Thread.sleep(forTimeInterval: 0.05)
        }
        let actual = currentCount()
        if actual > expected {
            XCTFail("Received more events than expected. Expected \(expected), got \(actual)", file: file, line: line)
        }
    }

    func validateNext(_ expected: EventValidation, file: StaticString = #file, line: UInt = #line) {
        lock.lock()
        let index = validationIndex
        let event = index < received.count ? received[index] : nil
        validationIndex += 1
        lock.unlock()

        guard let event = event else {
            XCTFail("No event at index \(index)", file: file, line: line)
            return
        }
        validate(event: event, against: expected, file: file, line: line)
    }

    private func currentCount() -> Int {
        lock.lock()
        defer { lock.unlock() }
        return received.count
    }

    private func validate(event: Event, against expected: EventValidation, file: StaticString, line: UInt) {
        switch expected {
        case .track(let name, let properties):
            guard let track = event as? TrackEvent else {
                XCTFail("Expected TrackEvent, got \(type(of: event))", file: file, line: line)
                return
            }
            XCTAssertEqual(track.event, name, "Track event name mismatch", file: file, line: line)
            assertJSONEqual(track.properties, properties, label: "Track properties", file: file, line: line)

        case .identify(let userId, let traits):
            guard let identify = event as? IdentifyEvent else {
                XCTFail("Expected IdentifyEvent, got \(type(of: event))", file: file, line: line)
                return
            }
            XCTAssertEqual(identify.userId, userId, "Identify userId mismatch", file: file, line: line)
            // Traits live under context["traits"] after the SDK processes the event.
            let contextTraits = identify.context?["traits"]?.value
            assertJSONEqual(contextTraits, traits, label: "Identify traits", file: file, line: line)

        case .screen(let screenName, let properties):
            guard let screen = event as? ScreenEvent else {
                XCTFail("Expected ScreenEvent, got \(type(of: event))", file: file, line: line)
                return
            }
            XCTAssertEqual(screen.event, screenName, "Screen event name mismatch", file: file, line: line)
            assertJSONEqual(screen.properties, properties, label: "Screen properties", file: file, line: line)

        case .group(let groupId, let traits):
            guard let group = event as? GroupEvent else {
                XCTFail("Expected GroupEvent, got \(type(of: event))", file: file, line: line)
                return
            }
            XCTAssertEqual(group.groupId, groupId, "Group groupId mismatch", file: file, line: line)
            assertJSONEqual(group.traits, traits, label: "Group traits", file: file, line: line)
        }
    }

    private func assertJSONEqual(_ received: Any?, _ expected: [String: Any], label: String, file: StaticString, line: UInt) {
        let receivedJSON = canonicalJSON(fromCodable: received)
        let expectedJSON = canonicalJSON(fromDictionary: expected)
        XCTAssertEqual(receivedJSON, expectedJSON, "\(label) mismatch", file: file, line: line)
    }

    private func canonicalJSON(fromCodable value: Any?) -> String {
        guard let value = value else { return "{}" }
        // If it's a CodableCollection, encode via JSONEncoder; otherwise assume [String: Any].
        if let collection = value as? CodableCollection {
            guard let data = try? JSONEncoder().encode(collection),
                  let obj = try? JSONSerialization.jsonObject(with: data),
                  let reencoded = try? JSONSerialization.data(withJSONObject: obj, options: [.sortedKeys]) else {
                return "<encode-failed>"
            }
            return String(data: reencoded, encoding: .utf8) ?? "<utf8-failed>"
        }
        if let dict = value as? [String: Any] {
            return canonicalJSON(fromDictionary: dict)
        }
        return "<unexpected-type: \(type(of: value))>"
    }

    private func canonicalJSON(fromDictionary dict: [String: Any]) -> String {
        guard let data = try? JSONSerialization.data(withJSONObject: dict, options: [.sortedKeys]),
              let str = String(data: data, encoding: .utf8) else {
            return "<serialize-failed>"
        }
        return str
    }
}
